src/lib/arch/riscv/InstructionMetadata.cc

```diff
@@ -0,0 +1,236 @@
+#include "InstructionMetadata.hh"
+
+#include <cassert>
+#include <cstring>
+#include <iostream>
+
+namespace simeng {
+namespace arch {
+namespace riscv {
+
+InstructionMetadata::InstructionMetadata(const cs_insn& insn)
+    : id(insn.id),
+      opcode(insn.opcode),
+      implicitSourceCount(insn.detail->regs_read_count),
+      implicitDestinationCount(insn.detail->regs_write_count),
+      //      groupCount(insn.detail->groups_count),
+      operandCount(insn.detail->riscv.op_count) {
+  std::memcpy(encoding, insn.bytes, sizeof(encoding));
+  // Copy printed output
+  std::strncpy(mnemonic, insn.mnemonic, CS_MNEMONIC_SIZE);
+  operandStr = std::string(insn.op_str);
+
+  // Copy register/group/operand information
+  std::memcpy(implicitSources, insn.detail->regs_read,
+              sizeof(uint16_t) * implicitSourceCount);
+  std::memcpy(implicitDestinations, insn.detail->regs_write,
+              sizeof(uint16_t) * implicitDestinationCount);
+  //  std::memcpy(groups, insn.detail->groups, sizeof(uint8_t) * groupCount);
+  std::memcpy(operands, insn.detail->riscv.operands,
+              sizeof(cs_riscv_op) * operandCount);
+
+  // Fix some inaccuracies in the decoded metadata
+  switch (opcode) {
+    case Opcode::RISCV_ADDI: {
+      if (operandCount == 0 && strcmp(mnemonic, "nop") == 0) {
+        operands[0].type = RISCV_OP_REG;
+        operands[0].reg = 1;
+
+        operands[1].type = RISCV_OP_REG;
+        operands[1].reg = 1;
+
+        operands[2].type = RISCV_OP_IMM;
+        operands[2].imm = 0;
+
+        operandCount = 3;
+      } else if (operandCount == 2 && strcmp(mnemonic, "mv") == 0) {
+        operands[2].type = RISCV_OP_IMM;
+        operands[2].imm = 0;
+
+        operandCount = 3;
+      }
+    }
+    case Opcode::RISCV_ADDIW: {
+      if (operandCount == 2 && strcmp(mnemonic, "sext.w") == 0) {
+        operands[2].type = RISCV_OP_IMM;
+        operands[2].imm = 0;
+
+        operandCount = 3;
+      }
+    }
+    case Opcode::RISCV_SUB: {
+      if (operandCount == 2 && strcmp(mnemonic, "neg") == 0) {
+        includeZeroRegisterPosOne();
+      }
+    }
+    case Opcode::RISCV_SUBW: {
```

Comment:
It's best to use `[[falltrough]]` to convey that the break statement is omitted intentionally for these cases.